### PR TITLE
chore: integrate pre-commit

### LINKED ISSUE #789: Integrate pre-commit
**Is your feature request related to a problem? Please describe.**
As a developer, I want other developers sending PRs to this project follow basic conventions used in this project

There are a couple of ways that such conventions can be enforced, but the most preferrred way is to have
pre-commit hooks that run in git before code is committed.

**Describe the solution you'd like**
- [ ] Research `pre-commit` and report what hooks can be used
- [ ] Implement those hooks
- [ ] Write Documentation

**Describe alternatives you've considered**
- Checking those conventions as part of CI
- Manually put instructions in README

**Additional context**

### PR DESCRIPTION
**Pull request by KushalBeniwal**
### Description

Continuing #1108 

Integrate pre-commit hook to automate running basic tools like black, flake8, isort before every commit message.

Fixes #789 

### Type of Change:

- Quality Assurance
- Documentation

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

with `python -m unittest discover tests`

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have made corresponding changes to the documentation
- [x] Update requirements.txt

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by KushalBeniwal** [comment]
@devkapilbansal except flake8 all pre-commit hooks pass for these changes.

**Comment by devkapilbansal** [comment]
@KushalBeniwal you can force push your changes too. No need to close and open a new PR

File: .flake8
Comment (devkapilbansal) [inline .flake8]: All of these erroors can be solved. Start solving first the errors you are getting by flake8 and then these too.

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
Great work @KushalBeniwal this PR is very close to be completed

**Comment by KushalBeniwal** [comment]
> @KushalBeniwal you can force push your changes too. No need to close and open a new PR

Ahh sorry. Out of the ways you suggested it looked like the fastest so I just did this.

**Comment by KushalBeniwal** [comment]
I have a hackathon this weekend. I'll complete this by Monday.
Sorry for any inconvenience caused.

**Comment by KushalBeniwal** [comment]
@devkapilbansal I resolved most of the errors from flake8.
The only ones remaining are:

`E501: line too long`

 This error is being generated for long strings and comments.
Comments I can work on but in strings dividing them in smaller substrings might make them harder to read and understand.
For eg: 

```
expected_result = "db_type_example://db_user_example:db_password_example@db_endpoint_example/db_name_example"
```

this can be divided like,
```
expected_result = ("db_type_example://"
                               "db_user_example:db_password_example@db_endpoint_example/"
                               "db_name_example")
```

But doing this makes it harder to read, but, if doing this is fine then I'll also resolve this error.

` C901 'validate_update_profile_request_data' is too complex`

In this a couple of errors are shown for a function name being too complex( > 18 chars). So how should I tackle this, as the meaning the name is conveying might get a little fuzzy if we shorten it.

`W503: line break before binary operator`

black wants it like this,

```
self.assertTrue(
    user.terms_and_conditions_checked
    == test_admin_user["terms_and_conditions_checked"]
)
```
 
and flake8 wants it like this,

```
self.assertTrue(
    User.terms_and_conditions_checked == 
    test_admin_user["terms_and_conditions_checked"]
)
```

if we do any of these the other gives error.

**Comment by devkapilbansal** [comment]
@KushalBeniwal you can increase the length of function name in `.flkae8` itself. 

* For the cases where black and flake8 are creating inconsistencies, I suggest using operators after line break as mentioned in [PEP8 guidelines](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).

* For the function is too complex, I would suggest ignore it for now and open an issue to simplify the function.

* What should we do with long comments can be discussed with @anitab-org/mentorship-maintainers. I am not sure if we should ignore this or splitting the line should be preferred.

Cc:- @isabelcosta

**Review by devkapilbansal — COMMENTED** [review]
Great work @KushalBeniwal 
See my comment above :point_up:

**Comment by KushalBeniwal** [comment]
@devkapilbansal I ignored these errors with descriptive comments. And now pre-commit runs successfully without throwing any errors.
I think it's almost done after we get other maintainers' views on what to do with comments.

**Comment by KushalBeniwal** [comment]
In addition to ignoring the errors should I also add max length for functions in flake8?

**Comment by devkapilbansal** [comment]
> In addition to ignoring the errors should I also add max length for functions in flake8?

I asked it in previous mentorship session. Will ask for a review again.

File: README.md
Comment (isabelcosta) [inline README.md]:
```suggestion
#### Automate auto-formatting tools with pre-commit
```

File: README.md
Comment (isabelcosta) [inline README.md]:
```suggestion
#### Sorting imports with isort
```

File: app/api/dao/mentorship_relation.py
Comment (isabelcosta) [inline app/api/dao/mentorship_relation.py:289]: wow nice catch, this was not being used at all 😱

File: run.py
Comment (isabelcosta) [inline run.py:17]: this was removed because it was not being used?

Comment (devkapilbansal) [reply run.py:17]: No this should not be removed. @KushalBeniwal please ignore it.

Comment (isabelcosta) [reply run.py:17]: thank you for checking @devkapilbansal 🙌🏾 I wasn't really sure. Everything else seems fine, I am ready to approve after this little change

**Comment by devkapilbansal** [comment]
> @devkapilbansal I ignored these errors with descriptive comments. And now pre-commit runs successfully without throwing any errors.
> I think it's almost done after we get other maintainers' views on what to do with comments.

Also, ensure in case of operator with line break, operator comes after line break

**Comment by KushalBeniwal** [comment]
> > @devkapilbansal I ignored these errors with descriptive comments. And now pre-commit runs successfully without throwing any errors.
> > I think it's almost done after we get other maintainers' views on what to do with comments.
> 
> Also, ensure in case of operator with line break, operator comes after line break

Yes @devkapilbansal I checked and operator comes after line break.

**Review by isabelcosta — APPROVED** [review]
Amazing work you did here @KushalBeniwal ! Both collaboration with maintainers and the contribution 💪🏾 
Thank you for contributing!

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
LGTM 👍 
Thanks @KushalBeniwal for making changes quickly and keeping this PR up to date

**Review by devkapilbansal — APPROVED** [review]

**Comment by devkapilbansal** [comment]
Accidentally clicked `request changes` earlier. 😬

**Comment by isabelcosta** [comment]
@devkapilbansal thank you so much for helping @KushalBeniwal here and pinging me after to help 🙌🏾 
I am tempted to make this a ready to merge PR, since the tests should catch anything that goes wrong with the API, also from what I have seen it's mostly style changes. I'll deploy this branch and just quickly check if I can log in and use a protected route. If everything fine, I'll merge :)

**Comment by isabelcosta** [comment]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses (positive and negative tests) were tested as below:

   * _Login and list all mentorship relations_  
     _Screenshot/gif_:  image
     _Expected Result_: API is working as expected
     _Actual Result_: Basic API functionalities are working

3. Additional Comments: I did a smoke test just to make sure the API was being loaded and functioning as expected. Most of the changes are only style-wise. I tested via deployment at https://ms-backend-review-pr-1117.herokuapp.com/

4. Status of PR Changed to: **Ready to Merge**

**Comment by isabelcosta** [comment]
@KushalBeniwal amazing work! I love how your commits are so specific and well divided with a specific change. Very well done 👏🏾

**Comment by KushalBeniwal** [comment]
Thank you @isabelcosta and @devkapilbansal for reviewing this PR!

**Pull request merged**
